CombinatorialGames/IGame/Short.lean

```diff
@@ -4,560 +4,175 @@ Released under Apache 2.0 license as described in the file LICENSE.
 
-/-- Perform a left move. -/
-def moveLeft : (x : SGame) → Fin x.leftMoves → SGame
-  | mk _ _ f _ => f
+theorem finite_rightMoves (x : IGame) [h : Short x] : x.rightMoves.Finite :=
+  (short_def.1 h).2.1
 
-/-- Perform a right move. -/
-def moveRight : (x : SGame) → Fin x.rightMoves → SGame
-  | mk _ _ _ g => g
+theorem finite_setOf_isOption (x : IGame) [Short x] : {y | IsOption y x}.Finite :=
+  (finite_leftMoves x).union (finite_rightMoves x)
 
-@[simp] theorem leftMoves_mk (m n f g) : (mk m n f g).leftMoves = m := rfl
-@[simp] theorem rightMoves_mk (m n f g) : (mk m n f g).rightMoves = n := rfl
-@[simp] theorem moveLeft_mk (m n f g) : (mk m n f g).moveLeft = f := rfl
-@[simp] theorem moveRight_mk (m n f g) : (mk m n f g).moveRight = g := rfl
+instance (x : IGame) [Short x] : Finite x.leftMoves :=
+  (Short.finite_leftMoves x).to_subtype
 
-/-- A well-founded relation on `SGame`, see `IGame.IsOption`. -/
-inductive IsOption : SGame → SGame → Prop
-  | moveLeft {x : SGame} (n : Fin x.leftMoves) : IsOption (x.moveLeft n) x
-  | moveRight {x : SGame} (n : Fin x.rightMoves) : IsOption (x.moveRight n) x
+instance (x : IGame) [Short x] : Finite x.rightMoves :=
+  (Short.finite_rightMoves x).to_subtype
 
-theorem isOption_wf : WellFounded IsOption := by
-  refine ⟨rec fun s t f g IHl IHr ↦ .intro _ ?_⟩
-  rintro y (h | h)
-  · exact IHl _
-  · exact IHr _
+protected theorem of_mem_leftMoves [h : Short x] (hy : y ∈ x.leftMoves) : Short y :=
+  (short_def.1 h).2.2.1 y hy
 
-instance : WellFoundedRelation SGame := ⟨_, isOption_wf⟩
+protected theorem of_mem_rightMoves [h : Short x] (hy : y ∈ x.rightMoves) : Short y :=
+  (short_def.1 h).2.2.2 y hy
 
-/-- See `igame_wf`. -/
-macro "sgame_wf" : tactic =>
-  `(tactic| all_goals solve_by_elim
-    [Prod.Lex.left, Prod.Lex.right, PSigma.Lex.left, PSigma.Lex.right,
-    IsOption.moveLeft, IsOption.moveRight] )
+protected theorem isOption [Short x] (h : IsOption y x) : Short y := by
+  cases h with
+  | inl h => exact .of_mem_leftMoves h
+  | inr h => exact .of_mem_rightMoves h
 
-def toString' (x : SGame) : String :=
-  "{" ++ ", ".intercalate (.ofFn fun i ↦ toString' (x.moveLeft i)) ++ " | " ++
-    ", ".intercalate (.ofFn fun j ↦ toString' (x.moveRight j)) ++ "}"
-termination_by x
-decreasing_by sgame_wf
+alias _root_.IGame.IsOption.short := Short.isOption
 
-/-- The size of an `SGame` is its number of proper subpositions.
-
-Note that this doesn't necessarily correspond to the number of subpositions of the `IGame`, though
-it does serve as a way to estimate how "complex" the game was to build. -/
-def size (x : SGame) : ℕ :=
-  ∑ i, ((x.moveLeft i).size + 1) + ∑ j, ((x.moveRight j).size + 1)
+theorem finite_setOf_subposition (x : IGame) [Short x] : {y | Subposition y x}.Finite := by
+  have : {y | Subposition y x} = {y | IsOption y x} ∪
+      ⋃ y ∈ {y | IsOption y x}, {z | Subposition z y} := by
+    ext
+    rw [Set.mem_setOf_eq, Subposition, Relation.transGen_iff]
+    simp [and_comm]
```

Comment:
(This feels like a good use case for suffices, but I couldn't get that to work.)